### PR TITLE
[DS-2940] Increase test coverage + fix Jupyter notebooks in Colab

### PR DESCRIPTION
**Pull request by nathancooperjones**
## Describe the big picture - why are we doing this?
Why have some tests when you can have many tests! 

In this PR, I went through missed lines in Codecov and tried to add as many tests as possible to Collie to cover those lines as well as a few extra edge cases I encountered. I time-boxed this ticket to two days of time, so it has been a painful two days writing tests, but look at this: **code coverage increased from 90% to 99%**!!! 

This PR also closes out https://github.com/ShopRunner/collie_recs/issues/7. 

Update: apparently I added 58 tests. Oof. 

## Any additional details to clarify code in the PR?
In addition to adding tests and fixing some edge cases, I also updated the notebooks to fully work in Colab by regenerating the data from the ``01`` notebook if needed. No code changes here really. 

I also standardized the header of all Jupyter notebooks to be very pretty! 

## How is this tested?
Haha, great joke. 

## Pull Request Checklist
 - [X] Pull request includes a description of why we are doing this
 - [x] ``CHANGELOG`` has been updated
 - [x] Version in ``_version.py`` has been updated
 - [X] All tests in the ``tests`` folder pass with a local build
 - [x] ``README`` has been updated (if applicable)
 - [ ] Documentation in ``docs`` has been updated (if applicable)
 - [X] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)

### CODE REVIEW

File: tests/test_cross_validation.py [hunk 1]
```diff
@@ -195,3 +195,68 @@ def test_stratified_split_with_user_with_only_one_interaction(
             test_p=0.2,
             seed=42,
         )
+
+
+def test_splits_with_wrong_p(interactions_to_split):
```
Comment (htorrence) [inline tests/test_cross_validation.py:200]: It's generally good software engineering practice to have each test only test a single action. When you chain together a bunch of different tests like this (conceptually related, but testing different bounds and multiple functions) you loose some clarity about what happened when the test fails. With this structure it's not immediately clear what caused the failure, and it's often helpful to know if you broke all the test splits, or just the stratified split ones, or just the upper bounds, etc, and this approach will only run the test to the first failure. If you want to have the logical groupings, how about putting them in a [test class](https://docs.pytest.org/en/stable/getting-started.html#group-multiple-tests-in-a-class)?

File: tests/fixtures/model_fixtures.py [hunk 2]
```diff
@@ -34,19 +34,43 @@ def implicit_model(train_val_implicit_data, gpu_count):
...
+                        'neucf_leaky_rulu',
+                        'neucf_custom',
+                        'hybrid_pretrained',
+                        'hybrid_pretrained_metadata_layers'])
+def models_trained_for_one_step(request,
+                                train_val_implicit_data,
+                                movielens_metadata_df,
+                                movielens_implicit_df,
+                                train_val_implicit_pandas_data,
+                                gpu_count):
```
Comment (htorrence) [inline tests/fixtures/model_fixtures.py:73]: Not technically part of this PR, but the use of tempfiles below to test the hdf5 work is really nice!

File: tutorials/02_matrix_factorization.ipynb [hunk 3]
```diff
@@ -1,671 +1,833 @@
...
-   "execution_count": 3,
-   "metadata": {},
-   "outputs": [
-    {
-     "name": "stderr",
-     "output_type": "stream",
-     "text": [
-      "Global seed set to 22\n"
-     ]
+  "nbformat": 4,
```
Comment (htorrence) [inline tutorials/02_matrix_factorization.ipynb:2]: Both reviewNB and github think most of this file changed, is it just the header or are there more actual changes to review in the notebooks?

**Review by htorrence — CHANGES_REQUESTED** [review]
Lots of good work here! Looks like the test-writing craze led to some nice fixes.

Some questions/comments, and I added some wording suggestions as well on errors messages - let me know if I've gotten the context wrong on any of those.

File: collie_recs/cross_validation.py [hunk 4]
```diff
@@ -63,7 +62,7 @@ def random_split(interactions: Interactions,
 
     """
     if len(kwargs) > 0 and [kwargs_key for kwargs_key in kwargs] != ['processes']:
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/cross_validation.py:64]: `kwargs` is still ignored, but I had a worry that if you accidentally passed an argument in wrong (like `testp` instead of `test_p`) that it would just silently work and use the default incorrectly. I added this code block to ensure that we check the `kwargs` are at least expected and, if not, throw an error that hopefully helps resolve this.

File: collie_recs/utils.py [hunk 5]
```diff
@@ -67,11 +67,6 @@ def _create_sparse_ratings_matrix_helper(users: Iterable[int],
                                          num_users: Union[int, str] = 'infer',
                                          num_items: Union[int, str] = 'infer') -> coo_matrix:
     """Create a sparse matrix from a series of arrays."""
-    # we found that putting this function in ``utils.py`` gave an error since ``utils.py`` imports
-    # ``interactions.py`` which imports ``utils.py`` and so on.
-    if min(users) != 0 or min(items) != 0:
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/utils.py:72]: Exactly!

File: tutorials/02_matrix_factorization.ipynb (hunk 3 above)
Comment (nathancooperjones) [reply tutorials/02_matrix_factorization.ipynb:2]: Just the header and the cell (cell 3) re-creating the data if not loaded from the file system. Everything else is exactly the same.

File: collie_recs/cross_validation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/cross_validation.py:64]: Nevermind, removed this!

File: collie_recs/model/hybrid_matrix_factorization_model.py [hunk 6]
```diff
@@ -336,13 +340,9 @@ def save_model(self,
         """
         path = str(path)
 
-        if os.path.isfile(path):
-            raise ValueError(f'``path`` must be a directory path! {path} not valid.')
-        elif os.path.exists(path):
+        if os.path.exists(path):
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/model/hybrid_matrix_factorization_model.py:343]: I realized the two checks were redundant, and I only needed a single check here!

File: tutorials/02_matrix_factorization.ipynb [hunk 7]
```diff
@@ -1,833 +1,803 @@
...
-      "pygments_lexer": "ipython3",
-      "version": "3.8.8"
-    },
-    "colab": {
-      "name": "02_matrix_factorization.ipynb",
-      "provenance": [],
-      "collapsed_sections": [
-        "44gLNrW-ghT_"
-      ]
+ "cells": [
```
Comment (nathancooperjones) [inline tutorials/02_matrix_factorization.ipynb:2]: Not sure why it is showing so much difference here, but the only thing I did was combine two cells together ¯\\_(ツ)_/¯

Comment (htorrence) [reply tutorials/02_matrix_factorization.ipynb:2]: Diffs on notebooks are weird. Thanks for clarifying!

**Review by htorrence — APPROVED** [review]
This looks great!
